public/app/features/explore/state/reducers.ts

```diff
@@ -591,12 +575,81 @@ export const itemReducer = reducerFactory<ExploreItemState>({} as ExploreItemSta
       const { loadingState } = action.payload;
       return {
         ...state,
-        loadingState,
+        queryResponse: {
+          ...state.queryResponse,
+          state: loadingState,
+        },
+        loading: loadingState === LoadingState.Loading || loadingState === LoadingState.Streaming,
       };
     },
   })
+  .addMapper({
+    //queryStreamUpdatedAction
+    filter: queryEndedAction,
+    mapper: (state, action): ExploreItemState => {
+      return processQueryResponse(state, action);
+    },
+  })
+  .addMapper({
+    filter: queryStreamUpdatedAction,
+    mapper: (state, action): ExploreItemState => {
+      return processQueryResponse(state, action);
+    },
+  })
   .create();
 
+export const processQueryResponse = (
+  state: ExploreItemState,
+  action: ActionOf<QueryEndedPayload>
+): ExploreItemState => {
+  const { response } = action.payload;
+  const { request, state: loadingState, series, legacy, error } = response;
+  const replacePreviousResults = action.type === queryEndedAction.type;
+
+  if (error) {
+    // For Angular editors
+    state.eventBridge.emit('data-error', error);
+
+    console.error(error); // To help finding problems with query syntax
+
+    if (!instanceOfDataQueryError(error)) {
+      response.error = toDataQueryError(error);
+    }
+
+    return {
+      ...state,
+      loading: false,
+      queryResponse: response,
+      graphResult: null,
+      tableResult: null,
+      logsResult: null,
+      showingStartPage: false,
+      update: makeInitialUpdateState(),
+    };
+  }
+
+  const latency = request.endTime - request.startTime;
+
+  // temporary hack until we switch to PanelData, Loki already converts to DataFrame so using legacy will destroy the format
+  const isLokiDataSource = state.datasourceInstance.meta.name === 'Loki';
+  const processor = new ResultProcessor(state, replacePreviousResults, isLokiDataSource ? series : legacy);
+
+  // For Angular editors
+  state.eventBridge.emit('data-received', processor.getRawData());
+
+  return {
+    ...state,
+    latency,
+    queryResponse: response,
+    graphResult: processor.getGraphResult(),
```

Comment:
I'm keeping the `graphResult`, `tableResult`and `logsResult`in state so that the UI components don't need to build it up with each render. These props are still in legacy format for this PR, next PR will change this too.